docs/enhancements/fep-001-api-fleet-templating.md

```diff
@@ -0,0 +1,295 @@
+As an IT administrator, I want to dial up the verbosity for logging on a specific device, so I can troubleshoot it.
+
+#### Story 7: region specific application service endpoints
+As an IT administrator, I want to provide region-specific service endpoints to my applications
+
+#### Story 8: region specific radio bands/regulations
+As a developer, I want to ensure that my devices automatically comply with local regulations when enrolled in specific regions. (i.e. configured with region-specific radio bands for WiFi, 5G, or other radio technologies)
+
+#### Story 9: region specific data sovereignty
+As a developer, I want to ensure that my devices connect to the specific regional endpoints to ensure compliance with data sovereignty.
+
+### Notes/Constraints/Caveats
+
+We would need to identify the design changes relative to templateVersions and
+if the relationship of those continues to be to the fleet or device level.
+
+#### Reference freezing drift
+
+One of the issues identified with this design, is that it's harder to pre-freeze the refences that would potentially
+surface from rendering all devices. For example if we had:
+
+```yaml
+os:
+  image: forklifts-{{ .device.metadata.label[factory] }}:latest
+```
+
+we couldn't know ahead of time what the image would be, and we would need to render all devices to know what the
+image would be. External git references with variable templating would have the same issue.
+
+Not being able to pre-freeze references while performing a roll-out means that some devices would potentially
+have one image, and if during rollout the image tag get's updated, the remaining devices will have a
+different one, while the referencing labels are still the same.
+
+##### Potential solutions
+
+1. We ignore it, and expect eventual consistency, on a 2nd round the server will detect that the image
+   reference has been updated, and the devices with the older images will be re-rendered.
+   Additionally, when using free-floating tags, it's probably a risk the customer is assuming in
+   favor of flexilibity and speed of development. This should in general be a practice we don't
+   recommend, as we only supporting it to make development easier while unfavouring determinism.
+
+2. We could mantain a list of all possible values for labels possibilities referenced within a fleet
+   This would help us create a list of references (i.e. an internal database object that holds all references)
+   for a rollout. This would not help with device names, but would help with labels.
+
+
+### Risks and Mitigations
+
+When rolling out a configuration to a fleet of devices, during the rollout, some devices could turn out to be
+not renderable, for example:
+* due to a mising label
+* because the template rendering is referencing non existing
+
+This could be mitigated by performing a dry-run of the rollout, and only applying the rollout if all devices
```

Comment:
We probably need to differentiate between "fleet controller did not find the label, so failed to do the substitution" and "the referenced resource (after substitution) does not exist / isn't accessible". There is little the service can do about the latter: verifying service-side is costly and also futile, because the resource could be there when the service tried, but no longer there when the agent needs it. At the end of the day, the agent will try during staging, fail, and set its condition to "failed to reconcile".